dspace-api/src/main/java/org/dspace/content/crosswalk/CrosswalkMetadataValidator.java

```diff
@@ -0,0 +1,115 @@
+
+import org.apache.commons.lang3.tuple.ImmutableTriple;
+import org.apache.commons.lang3.tuple.Triple;
+import org.dspace.authorize.AuthorizeException;
+import org.dspace.content.MetadataField;
+import org.dspace.content.MetadataSchema;
+import org.dspace.content.NonUniqueMetadataException;
+import org.dspace.content.factory.ContentServiceFactory;
+import org.dspace.content.service.MetadataFieldService;
+import org.dspace.content.service.MetadataSchemaService;
+import org.dspace.core.ConfigurationManager;
+import org.dspace.core.Context;
+
+import java.sql.SQLException;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.Map;
+
+public class CrosswalkMetadataValidator {
+
+    protected MetadataSchemaService metadataSchemaService;
+    protected MetadataFieldService metadataFieldService;
+
+    private String schemaChoice;
+    private String fieldChoice;
+
+    private Map<Triple<String, String, String>, MetadataField> validatedMetadataFields;
+
+    public CrosswalkMetadataValidator() {
+        metadataSchemaService = ContentServiceFactory.getInstance().getMetadataSchemaService();
+        metadataFieldService = ContentServiceFactory.getInstance().getMetadataFieldService();
+
+        validatedMetadataFields = new HashMap<>();
+
+        // The two options, with three possibilities each: add, ignore, fail
+        schemaChoice = ConfigurationManager.getProperty("oai", "harvester.unknownSchema");
+        if (schemaChoice == null)
+        {
+            schemaChoice = "fail";
+        }
+
+        fieldChoice = ConfigurationManager.getProperty("oai", "harvester.unknownField");
+        if (fieldChoice == null)
+        {
+            fieldChoice = "fail";
+        }
+    }
+
+    /**
+     * Scans metadata for elements not defined in this DSpace instance. It then takes action based
+     * on a configurable parameter (fail, ignore, add).
+     */
+    public MetadataField checkMetadata(Context context, String schema, String element, String qualifier, boolean forceCreate) throws SQLException, AuthorizeException, CrosswalkException {
+        if(!validatedBefore(schema, element, qualifier)) {
+            // Verify that the schema exists
+            MetadataSchema mdSchema = metadataSchemaService.find(context, schema);
+            MetadataField mdField = null;
+
+            if (mdSchema == null) {
+                // add a new schema, giving it a namespace of "unknown". Possibly a very bad idea.
+                if (forceCreate && schemaChoice.equals("add")) {
+                    try {
+                        mdSchema = metadataSchemaService.create(context, schema, String.valueOf(new Date().getTime()));
+                        mdSchema.setNamespace("unknown" + mdSchema.getID());
+                        metadataSchemaService.update(context, mdSchema);
+                    } catch (NonUniqueMetadataException e) {
+                        // This case should not be possible
+                        e.printStackTrace();
```

Comment:
(1) User is not told why he is seeing a stack trace.  (2) Error output should go to System.err:  e.printStackTrace(System.err);